### ISSUE TITLE
New Salmon indices with Genome

### ISSUE THREAD

**Issue by k3yavi** [databio/refgenieserver#63]
Hi guys,

Thanks for maintaining such an awesome resource.
I am from the salmon developer's group and we recently released a [new version](https://github.com/COMBINE-lab/salmon/releases/tag/v1.0.0) of salmon that supports highly accurate quantification (check [this](https://www.biorxiv.org/content/10.1101/657874v2) preprint for more details). Some of our users are interested in pre built salmon indices https://github.com/COMBINE-lab/salmon/issues/444 and we are exploring the resources to upload new Genome based salmon indices which are downloadable and queryable, may be based on species or the reference sequence. I found the refgenie very useful and am wondering are you guys open to updating your database with  the latest salmon indices ? For starters support for just the human and mouse indices would be great too.

**Comment by stolarczyk** [comment]
Thanks!
Sure, we're willing to update the resources. We can easily build and serve `salmon_index` assets produced by the new `salmon` version, but having a containerized version of `salmon` here: https://biocontainers.pro/#/tools/salmon would allow others to easily build assets using `bulker`: http://bulker.databio.org/en/latest/refgenie_tutorial/. Could you reach out to them and ask to containerize the newest `salmon` release?

The new version of `salmon_index` assets will be up in the next few days.

**Comment by stolarczyk** [comment]
> Could you reach out to them and ask to containerize the newest `salmon` release?

nvm, I just realized the new version is already available

**Comment by stolarczyk** [comment]
@k3yavi, `salmon_index` assets are up at: http://refgenomes.databio.org/

However, even though the build configurations remained the same, the indexes are generally much smaller (hundreds of MB vs. >2GB) than the ones produced with `salmon v0.14.2`. Is that expectable?

Let's take `mm10` genome as an example if you wish to inspect the build process:
build log: http://refgenomes.databio.org/v2/asset/mm10_cdna/salmon_index/log?tag=default
build recipe: http://refgenomes.databio.org/v2/asset/mm10_cdna/salmon_index/recipe?tag=default

**Comment by k3yavi** [comment]
HI @michalstolarczyk ,

Thanks for the swift replies and rapid updates !
The low index size is indeed an expected behavior, in fact, it enables salmon to index the full genome now. With the latest release we proposed three different way of indexing salmon:

(In the order of decreasing  quantification accuracy; decreasing index size, decreasing running time)
1.) We call Selective Alignment Full (SAF): If you follow [this](https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment/) tutorial, you can index both the genome and transcriptome with Salmon. The caveat being index size for human can be ~20G .
2.) It's a layman version of Selective Alignment. Basically we mapped txome to genome and extracted the relevant portion out from the genome and index it along with the transcriptome. Currently, we are maintaining these extra sequence for ~20 model organisms for users [here](http://bit.ly/30yn3FJ). The caveat here is the quantification estimates through this indexing are definitely better than just using cdna (txome) indexing but worse than full genome indexing. It's suggested for users with less RAM and it takes ~2G .
3) Is the traditional way of indexing just the transcriptome. It's generates the least accurate quantification estimates however takes like total of ~500Mb memory.

I hope this makes sense and would love to hear back about which versions (1,2,3, all) you'd like to upload into refgenie. Also I was wondering your thought about extending the framework for other model organisms like rat etc.

**Comment by nsheff** [comment]
This all sounds doable. Each of these are different assets, with different recipes. We have already done number 3. 

Number 1 and 2 are possible, but will require some updates to refgenie, because we haven't encountered the idea of an asset that requires 2 parents of the same type (in this case, 2 parent fasta files). But, we should accommodate this. We have figured out exactly what we have to do, which are actually just a few small updates, for these new features in refgenie:

https://github.com/databio/refgenie/issues/139
https://github.com/databio/refgenie/issues/140

Once we have that set up...we can easily do number 1, because we can follow your tutorial. To do number 2, we would need you to provide a **recipe** that can produce this index, given any inputs.

**Comment by nsheff** [comment]
> Also I was wondering your thought about extending the framework for other model organisms like rat etc.

We're open to other common model organisms -- we do have rat on there already, so we are set up to do human, mouse, and rat right now. which others are you thinking of?

**Comment by k3yavi** [comment]
Hi @nsheff ,

Thanks for the speedy integration of our requests and it was great meeting you in person at the CSHL meeting. We are excited to use and refer salmon users to refgenie's automated pipeline. Once refgenie is ready with the new indexing (scheme 1), I think it's worth adding a short tutorial for end-users. We are maintaining the tutorials [here](https://combine-lab.github.io/alevin-tutorial) and I am happy to help in that respect as well. 

To answer your questions about scheme 2, we used [this](https://github.com/COMBINE-lab/SalmonTools/blob/master/scripts/generateDecoyTranscriptome.sh) script for generating the decoys, however, mashmap (we are using it for mapping genome to transcriptome) can take significantly high memory (~100G for human). In lieu of that we have prepared the require files for salmon indexing on our server and shared it online for ~20 model organisms [here](https://drive.google.com/drive/u/1/folders/1MeM5BCyUjPD9UTzRUJpAWgI_q4zJj42n). Basically each zipped file in the google drive is named after the species and contains two files `decoys.txt` and `gentrome.fa.gz`. Once the files are available for the relevant species, the recipe for salmon indexing (scheme 2) is as easy as doing 
`salmon index -t gentrome.fa.gz -d decoys.txt -i <index_directory> -p <#threads>`.

I think starting with human, mouse and rat is perfect. Looping in @rob-p .
I was discussing this with Rob earlier and we thought may be majority vote on the most used model organisms would be a good idea. How about we loop you guys in once we have more information regarding this and/or a user request the index for some new species ?

**Comment by nsheff** [comment]
Hi @k3yavi, we're finished in principle with the updates required to accommodate this new type of salmon index. We'll probably release a new version in the next few days with this capability. 

So we're now ready for a list of specific assets to build --  as in, which exact txome and genome inputs for which versions of the human genome, etc.  It would be helpful if you could get us started with a list of how you envision this. We can try to come to a consensus with what we have.

I propose we create a github repository to track the list of inputs and assets we will build. Adding a new species or version then would be done via a PR to this repository. We have a structure we are using in these two example repositories: 

https://github.com/databio/databio_genomes
https://github.com/databio/refgenie_bowtie

Basically, we would need a `genomes.tsv` file that specifies the input files (these could be URLs to remotely hosted fasta files, for example). It would look something like this:

sample_name | txome_fasta | genome_fasta
------ | ----- | -----
hg38 | path/to/txome.fa | path/to/genome.fa
hg19 | path/to/txome.fa | path/to/genome.fa

Then we need an `assets.tsv` file with the asset name to build for each genome

sample_name | asset 
------ | ----- 
hg38 | genome_fasta
hg38 | txome_fasta
hg38 | salmon_index
hg19 | genome_fasta
hg19 | txome_fasta
hg19 | salmon_index

The salmon index assets will be built from the parent fasta assets, which allows us to keep track of how things were built. You can see what this looks like for current salmon assets on our [staging server](http://staging.refgenomes.databio.org/v2/asset/hg19_cdna/salmon_index/splash?tag=default). The new salmon indexes would have 2 parent assets. I think the names aren't quite finalized and we can probably handle the `assets.tsv` file.

So, what I think would be useful for us at this point is if you can start to put together your vision for the actual inputs you're interested in for the `genomes.tsv` file.

**Comment by k3yavi** [comment]
Hi @nsheff ,

Thanks again for the heads up and updating refgenie for the new requirements of salmon indexing.
I think I like the idea of maintaining a github repo for metadata information about assets into a  `genomes.tsv` file. It's clean and help us redirect other people in case they have specific requirements, obviously it would be under your discretion to what extend you'd like to maintain the requests. 
From our side, to get the ball rolling if you can share what repo where I should make the PR for the `genomes.tsv` file with ftp link, that'd be great. I am thinking of hg38, mm10 and rat ensembl reference, I can also drop the tsv file here on the chat if you'd prefer.

**Comment by stolarczyk** [comment]
@k3yavi ,
We've run into a problem building `salmon_sa_index`es (following [this](https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment/) tutorial).
We got the error below for every genome/txome we tried: `mm10`, `hg38`, `hg19` and `rn6`

```
salmon: line 5: 24565 Segmentation fault      (core dumped)
```

couple of facts: 
- memory limit should not be an issue since 60GB was requested
- we're running `salmon` with singularity using this container: quay.io/biocontainers/salmon:1.0.0--hf69c8f4_0

log files:
- [commands](https://github.com/databio/refgenieserver/files/3918103/refgenie_commands.log)
- [build log](https://github.com/databio/refgenieserver/files/3918104/refgenie_log.log)

Have you encountered this issue before?

**Comment by k3yavi** [comment]
Hi @michalstolarczyk ,

Thanks for reporting, we have not seen this error before. I'll try to use the singularity container. I can grab the mm10 reference from Ensembl, can you confirm that you are also using the reference off the shelf from Ensembl ?

**Comment by stolarczyk** [comment]
For mouse we used `mm10/fasta:default` and  `mm10_cdna/fasta:default` assets from refgenie:

You can easily get them with:
```
refgenie init -c genomes.yaml
refgenie pull mm10/fasta:default -c genomes.yaml
refgenie pull mm10_cdna/fasta:default -c genomes.yaml
```

or just download from refgenieserver: 
- http://staging.refgenomes.databio.org/v2/asset/mm10/fasta/archive?tag=default
- http://staging.refgenomes.databio.org/v2/asset/mm10_cdna/fasta/archive?tag=default

these are the sources we got them from:
- ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.5_GRCm38.p3/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001635.5_GRCm38.p3_no_alt_analysis_set.fna.gz
- ftp://ftp.ensembl.org/pub/release-97/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz

**Comment by k3yavi** [comment]
Cool, thanks, let me try and replicate the error on my end. I will get back to you in a bit.

**Comment by k3yavi** [comment]
Hi @michalstolarczyk ,

I extracted the files from the refgenie link you shared above.
I am not sure why but both files starts from weird binary text after unzipping the fasta file, is it expected?
```
fasta/@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@0002775@3420401@0000144@00000000000@13552377627@011316@ 5@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ustar  @mjs5kd
@@@@@@@@@@@@@@@@@@@@@@@@@@users@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@fasta/mm10_cdna.fa@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@0000664@3420401@0000144@01677751502@13551640373@013410@ 0@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ustar  @mjs5kd@@@@@@@@@@@@@@@@@@@@@@@@@@user
s@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
```
I am gonna go ahead with the ncbi ftp link to check for the segfault.

**Comment by stolarczyk** [comment]
the assets on the server are tarred and gzipped, hence `.tgz` extension. Did you untar it? Also there are 3 files in every `fasta` asset tarball: .fa, .fai and chrom.sizes.

running `refgenie pull mm10/fasta:default -c genomes.yaml` does it for you.

```
$ head -2 `refgenie seek mm10_cdna/fasta -c genomes.yaml`
>ENSMUST00000196221.1 cdna chromosome:GRCm38:14:54113468:54113476:1 gene:ENSMUSG00000096749.2 gene_biotype:TR_D_gene transcript_biotype:TR_D_gene gene_symbol:Trdd1 description:T cell receptor delta diversity 1 [Source:MGI Symbol;Acc:MGI:4439547]
ATGGCATAT
```

**Comment by k3yavi** [comment]
duh ! My bad, there was no extensions to the file when downloaded with wget so I assumed just gzipped. Thanks for the heads up.

**Comment by k3yavi** [comment]
@michalstolarczyk , I can confirm that salmon finishes indexing successfully on our end using the singularity docker. I missed to profile the run, restarted again for the stats on memory usage.

Any idea what else can be wrong ?

**Comment by nsheff** [comment]
Interesting -- were you using singularity or docker? can you provide the shell command you used successfully?

**Comment by k3yavi** [comment]
Made some progress, I think it's the RAM. ~Seemingly when I use the Ensembl reference I use 14G while~ when I use the refgenie reference I use 75G -- I think which is why the segfault.
I think there is some corner case in processing the files with salmon, I'll dig more and keep you guys updated,

**Correction:** It was taking 75G for both the reference on docker

**Comment by k3yavi** [comment]
@nsheff Saw your comment late, I tried both but with docker I am using: 
`docker run -it quay.io/biocontainers/salmon:1.0.0--hf69c8f4_0`
and the salmon commands are the same as forwarded by Michal above.

**Comment by nsheff** [comment]
> `docker run -it quay.io/biocontainers/salmon:1.0.0--hf69c8f4_0`
> and the salmon commands are the same as forwarded by Michal above.

ok perfect, I'm testing it locally now. But if it takes 75G then that's likely the issue. @michalstolarczyk can you try again requesting more memory on the hpc? my desktop isn't going to cut it...

Also, I wonder why our peak memory calculation was off, since pypiper reported only 20G used...

**Comment by k3yavi** [comment]
Thanks @nsheff for all  the help. We are working on checking it on a local comp as well because ideally we don;t expect salmon to take that much memory, one hypothesis was may be docker is exacerbating the memory usage somehow, but not sure right now. In case we are checking on the HPC it's worth putting a little higher than 75 as it was 75.7 ish.

**Comment by stolarczyk** [comment]
I'm running two jobs now:

- `refgenie build ...` (which is using refgenie assets as inputs) with the memory limit of 100GB in a container managed by `bulker` (singularity)
- manual `salmon index ...` with the references retrieved from the ftp sources with 60GB memory limit in a container managed by `bulker` (singularity)

**Comment by k3yavi** [comment]
Thanks @michalstolarczyk ,
I just finished indexing on my local comp without docker. I can finish indexing using the same reference in `<20G`. Please excuse my systems knowledge, does the virtual memory is also counted as regular memory by the software on the HPC ? Because that might explain it why profiler was saying 19G but the program segfaulted due to high virtual memory usage.

A correction on the previous text, it was using 75G on both the enseml and refgenie reference when profiled from within the docker.

**Comment by stolarczyk** [comment]
in `pypiper` we just record the _physical_ memory with [`psutil.Process.memory_info().rss`](https://psutil.readthedocs.io/en/latest/index.html#psutil.Process.memory_info)

do you suspect the virtual memory to be used here? even though there's plenty of RAM allocated for the job?

**Comment by k3yavi** [comment]
It's possible but I am not completely sure as the psutils seem to have another variable for monitoring the virtual memory. I am not sure what else could be different in the docker environment which is making salmon use excess memory. 

This is how the time looks on our server from inside docker
 ```
         Command being timed: "salmon index -t gentrome.fa -d decoys.txt -p 20 -i index"
         User time (seconds): 7338.52
         System time (seconds): 118.91
         Percent of CPU this job got: 592%
         Elapsed (wall clock) time (h:mm:ss or m:ss): 20m 57.77s
         Average shared text size (kbytes): 0
         Average unshared data size (kbytes): 0
         Average stack size (kbytes): 0
         Average total size (kbytes): 0
         Maximum resident set size (kbytes): 75740864
         Average resident set size (kbytes): 0
         Major (requiring I/O) page faults: 0
         Minor (reclaiming a frame) page faults: 3654794
         Voluntary context switches: 837014
         Involuntary context switches: 31020
         Swaps: 0
         File system inputs: 0
         File system outputs: 94852408
         Socket messages sent: 0
         Socket messages received: 0
         Signals delivered: 0
         Page size (bytes): 4096
         Exit status: 
```

**Comment by k3yavi** [comment]
One more interesting observation, If I use the salmon docker image from Combine-Lab 
`docker run -it combinelab/salmon` I can still index with `<20G` RAM. I think this narrows down the problem to the salmon image used by biocontainers peeps which is messing up with salmon's memory usage because if profiling was the issue then it should show similar memory usage with combine lab docker image.

> refgenie build ... (which is using refgenie assets as inputs) with the memory limit of 100GB in a container managed by bulker (singularity)

Curious if  ⬆️ was able to finish successfully ?

**Comment by nsheff** [comment]
I ran it successfully locally using docker and the memory use reported by docker stats maxed out around 24Gb.

**Comment by stolarczyk** [comment]
and you ran in with refgenie assets, right?

**Comment by nsheff** [comment]
yeah. My command was this: 

```
docker run -it -v /ext:/ext -v /project:/project \
quay.io/biocontainers/salmon:1.0.0--hf69c8f4_0 \
salmon index \
-t /project/shefflab/genomes/hg38/salmon_sa_index/default/gentrome.fa \
-d /project/shefflab/genomes/hg38/salmon_sa_index/default/decoys.txt \
-p 8 -i /project/shefflab/genomes/hg38/salmon_sa_index/ns5bc_test
```

**Comment by stolarczyk** [comment]
couple of updates:

- I run the indexer with the same inputs as @nsheff  above as `refgenie build hg38/salmon_sa_index --assets genomefa=hg38/fasta txomefa=hg38_cdna/fasta` using the `combine/salmon` container with singularity and I got the same error. Memory limit was 30Gb, which should be enough considering previous memory usage. The build failed with the segmentation error.
- to rule out the possibility that refgenie is the source of the problem I run the same command not from refgenie but using singularity as the container engine (on the same input files). Memory limit was 30Gb and the container was `combine/salmon` again.

command:
```
singularity exec -B /project \
 /home/mjs5kd/simages/combinelab/salmon:1.0.0 salmon index \
 -t /project/shefflab/genomes/hg38/salmon_sa_index/default/gentrome.fa \
 -d /project/shefflab/genomes/hg38/salmon_sa_index/default/decoys.txt -p 8 \
 -i  /project/shefflab/genomes/hg38/salmon_sa_index/mjs5kd_test
```
 but I got the `Segmentation fault (core dumped)` error
 
- to rule out that the problem was with the `combine/salmon` container I run the same command in the same container. The only difference this time was the container engine -- singularity. The memory limit was still 30Gb. Consequently, the only difference between @nsheff's and this command was the container engine and the fact that I run it on HPC, not locally.

command:
```
singularity exec -B /project \
 /home/mjs5kd/simages/quay.io/biocontainers/salmon:1.0.0--hf69c8f4_0 salmon index \ 
 -t /project/shefflab/genomes/hg38/salmon_sa_index/default/gentrome.fa \
 -d /project/shefflab/genomes/hg38/salmon_sa_index/default/decoys.txt -p 8 \
 -i  /project/shefflab/genomes/hg38/salmon_sa_index/mjs5kd_test
```

and the segmentation error still occurred

**Comment by k3yavi** [comment]
Thanks @michalstolarczyk for the extensive testing. The fact that it works locally but not on the HPC makes me wonder if there is some restrictions on HPC for virtual memory which we are not aware of ? Like can we check / increase the virtual memory limit, if possible? I am not a systems person please excuse if I talking stupid.

**Comment by nsheff** [comment]
I indexed it successfully using `bulker/singularity` and `refgenie build` on my local computer. So I believe it's a problem on our HPC. I will now try to see if it works for me using singularity on HPC.

**Comment by nsheff** [comment]
I reproduced the segmentation fault on the HPC with identical command and identical inputs. I don't think it's memory use. The software is outputting different things.

I notice the results are identical until a certain point.

In the working command log it says: 

```
Round 0, 0:68719476736
Pass	Filling	Filtering
1	642	2785	
2	754	8
True junctions count = 22106118
False junctions count = 3915253
Hash table size = 26021371
Candidate marks count = 351646420
--------------------------------------------------------------------------------
Reallocating bifurcations time: 17
True marks count: 347154688
Edges construction time: 1697
--------------------------------------------------------------------------------
Distinct junctions = 22106118
```
The failed command log says; 
```
Round 0, 0:68719476736
Pass	Filling	Filtering
1	190	918	
2	63	9
True junctions count = 22106118
False junctions count = 3575582
Hash table size = 25681700
Candidate marks count = 351260932
--------------------------------------------------------------------------------
Reallocating bifurcations time: 14
True marks count: 347154688
Edges construction time: 486
--------------------------------------------------------------------------------
Distinct junctions = 22106118
```



@k3yavi can you point me to what in salmon might lead to different results at this point?


Here's the complete working version:
```
Threads = 8
Vertex length = 31
Hash functions = 5
Filter size = 68719476736
Capacity = 2
Files: 
/ext/qumulo/sandbox/refgenie_test/hg38/salmon_sa_index/default/ref_k31_fixed.fa
--------------------------------------------------------------------------------
Round 0, 0:68719476736
Pass	Filling	Filtering
1	642	2785	
2	754	8
True junctions count = 22106118
False junctions count = 3915253
Hash table size = 26021371
Candidate marks count = 351646420
--------------------------------------------------------------------------------
Reallocating bifurcations time: 17
True marks count: 347154688
Edges construction time: 1697
--------------------------------------------------------------------------------
Distinct junctions = 22106118

approximateContigTotalLength: 1298055990
counters:
5265582 1149 1192 70
contig count: 36981178 element count: 3784352032 complex nodes: 5267993
size: 3784352032
# of ones in rank vector: 36981177
size: 3784352032
[2019-12-04 18:46:07.192] [puff::index::jointLog] [info] Setting the index/BinaryGfa directory /ext/qumulo/sandbox/refgenie_test/hg38/salmon_sa_index/default
size = 3784352032
-----------------------------------------
| Loading contigs | Time = 24.648 s
-----------------------------------------
size = 3784352032
-----------------------------------------
| Loading contig boundaries | Time = 12.75 s
-----------------------------------------
Number of ones: 36981177
Number of ones per inventory item: 512
Inventory entries filled: 72229
[2019-12-04 18:46:50.141] [puff::index::jointLog] [info] Done wrapping the rank vector with a rank9sel structure.
[2019-12-04 18:47:06.768] [puff::index::jointLog] [info] contig count for validation: 36981177
[2019-12-04 18:48:33.107] [puff::index::jointLog] [info] Total # of Contigs : 36981177
[2019-12-04 18:48:33.107] [puff::index::jointLog] [info] Total # of numerical Contigs : 36981177
[2019-12-04 18:50:39.954] [puff::index::jointLog] [info] 
Total # of segments we have position for : 36981177
[2019-12-04 18:50:41.365] [puff::index::jointLog] [info] total contig vec entries 360587850
[2019-12-04 18:50:41.371] [puff::index::jointLog] [info] bits per offset entry 29
[2019-12-04 18:51:09.010] [puff::index::jointLog] [info] there were 3457714  equivalence classes
[2019-12-04 18:52:17.559] [puff::index::jointLog] [info] # segments = 36981177
[2019-12-04 18:52:17.559] [puff::index::jointLog] [info] total length = 3784352032
[2019-12-04 18:52:17.903] [puff::index::jointLog] [info] Reading the reference files ...
[2019-12-04 18:53:33.042] [puff::index::jointLog] [info] positional integer width = 32
[2019-12-04 18:53:33.042] [puff::index::jointLog] [info] seqSize = 3784352032
[2019-12-04 18:53:33.042] [puff::index::jointLog] [info] rankSize = 3784352032
[2019-12-04 18:53:33.042] [puff::index::jointLog] [info] edgeVecSize = 0
[2019-12-04 18:53:33.042] [puff::index::jointLog] [info] num keys = 2674916722
[Building BooPHF]  100  %   elapsed:   5 min 37 sec   remaining:   0 min 0  sec
```

The failed version: 
```
Threads = 8
Vertex length = 31
Hash functions = 5
Filter size = 68719476736
Capacity = 2
Files: 
/project/shefflab/sandbox/refgenie_test/hg38/salmon_sa_index_hpc/default/ref_k31_fixed.fa
--------------------------------------------------------------------------------
Round 0, 0:68719476736
Pass	Filling	Filtering
1	190	918	
2	63	9
True junctions count = 22106118
False junctions count = 3575582
Hash table size = 25681700
Candidate marks count = 351260932
--------------------------------------------------------------------------------
Reallocating bifurcations time: 14
True marks count: 347154688
Edges construction time: 486
--------------------------------------------------------------------------------
Distinct junctions = 22106118

approximateContigTotalLength: 1298055990
counters:
5265582 1161 1180 70
contig count: 36981178 element count: 3784352032 complex nodes: 5267993
size: 3784352032
# of ones in rank vector: 36981177
size: 3784352032
[2019-12-05 08:51:47.626] [puff::index::jointLog] [info] Setting the index/BinaryGfa directory /project/shefflab/sandbox/refgenie_test/hg38/salmon_sa_index_hpc/default
size = 3784352032
-----------------------------------------
| Loading contigs | Time = 530.94 ms
-----------------------------------------
size = 3784352032
-----------------------------------------
| Loading contig boundaries | Time = 263.61 ms
-----------------------------------------
Number of ones: 36981177
Number of ones per inventory item: 512
Inventory entries filled: 72229
[2019-12-05 08:51:53.737] [puff::index::jointLog] [info] Done wrapping the rank vector with a rank9sel structure.
[2019-12-05 08:52:10.891] [puff::index::jointLog] [info] contig count for validation: 36981177
[2019-12-05 08:52:46.684] [puff::index::jointLog] [info] Total # of Contigs : 36981177
[2019-12-05 08:52:46.684] [puff::index::jointLog] [info] Total # of numerical Contigs : 36981177
[2019-12-05 08:54:42.734] [puff::index::jointLog] [info] 
Total # of segments we have position for : 36981177
[2019-12-05 08:54:44.346] [puff::index::jointLog] [info] total contig vec entries 360587850
[2019-12-05 08:54:44.347] [puff::index::jointLog] [info] bits per offset entry 29
[2019-12-05 08:55:11.418] [puff::index::jointLog] [info] there were 3457714  equivalence classes
[2019-12-05 08:55:50.786] [puff::index::jointLog] [info] # segments = 36981177
[2019-12-05 08:55:50.786] [puff::index::jointLog] [info] total length = 3784352032
[2019-12-05 08:55:51.310] [puff::index::jointLog] [info] Reading the reference files ...
[2019-12-05 08:56:14.917] [puff::index::jointLog] [info] positional integer width = 32
[2019-12-05 08:56:14.917] [puff::index::jointLog] [info] seqSize = 3784352032
[2019-12-05 08:56:14.917] [puff::index::jointLog] [info] rankSize = 3784352032
[2019-12-05 08:56:14.917] [puff::index::jointLog] [info] edgeVecSize = 0
[2019-12-05 08:56:14.917] [puff::index::jointLog] [info] num keys = 2674916722
/project/shefflab/bulker/bulker_crates/databio/refgenie/0.7.3/salmon: line 6: 193122 Segmentation fault      (core dumped) LC_ALL=C singularity exec -B "/tmp:/tmp" -B "/sfs:/sfs" /project/shefflab/bulker/simages/combinelab/salmon:1.0.0 salmon "$@"
```

**Comment by k3yavi** [comment]
Hi @nsheff ,

Thanks for sharing the logs. I think the logs are producing the same output.
The section of log you shared is for the de-bruijn graph generation phase and uses bloom filters which can results in different number of FP between the two runs.
After the debruij is created, the total number of contigs is same i.e.
```
contig count: 36981178 element count: 3784352032 complex nodes: 5267993
```
are same in both the logs and seem to index until the point of the creating the minimum perfect hash. Seemingly `num keys = 2674916722` for creating the minimum perfect hash is also same and somehow one process segfaults just before starting to create the MPH.

**Comment by nsheff** [comment]
Maybe related to: https://github.com/COMBINE-lab/salmon/issues/271 ?

**Comment by nsheff** [comment]
Our HPC team has discovered that the segfault occurs when using salmon installed natively, not only on singularity. So, it is not related to singularity but may be related to SLURM. Have you run salmon successfully using slurm?

**Comment by k3yavi** [comment]
We do have a HPC system at University level which uses SLURM but we rarely use, I'll try indexing salmon using that now.

**Comment by nsheff** [comment]
> Hi @nsheff ,
> 
> Thanks again for the heads up and updating refgenie for the new requirements of salmon indexing.
> I think I like the idea of maintaining a github repo for metadata information about assets into a `genomes.tsv` file. It's clean and help us redirect other people in case they have specific requirements, obviously it would be under your discretion to what extend you'd like to maintain the requests.
> From our side, to get the ball rolling if you can share what repo where I should make the PR for the `genomes.tsv` file with ftp link, that'd be great. I am thinking of hg38, mm10 and rat ensembl reference, I can also drop the tsv file here on the chat if you'd prefer.

Ok back to the original thread topic; we're ready with this now.  [Here is the repository](https://github.com/databio/databio_genomes) where we're cataloging the assets and all their inputs. I have added some instructions on how to add a new asset to this here:

https://github.com/databio/databio_genomes/blob/master/asset_pep/README.md#adding-an-asset-to-this-pep

It looks like @michalstolarczyk has already added a few of the basic salmon indexes (the combined genome/txome ones) to this. So, can you see if those instructions make sense, and take a look at the examples? Then, feel free to add some new versions you're interested in via a PR to that repository.

Let me know if something is unclear and I will try to clarify the docs. This is the first time doing this so there may be some things I need to document better.

**Comment by k3yavi** [comment]
I can confirm the indexing finished successfully using slurm on our HPC .

@nsheff  thanks for the forwarding the docs for adding the PR with the request for new assets.
It looks straightforward and makes sense.

**Comment by stolarczyk** [comment]
FYI, the salmon recipe names are:
- `salmon_index`: the basic transcriptome index, so it belongs to `*_cdna` namespaces, e.g `hg38_cdna`
- `salmon_sa_index`: Selective Alignment Full (SAF) version, based on [this tutorial](https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment)
- `salmon_partial_sa_index`: layman Selective Alignment version, based on [this script]( https://github.com/COMBINE-lab/SalmonTools/blob/master/scripts/generateDecoyTranscriptome.sh)

the recipes are defined [here](https://github.com/databio/refgenie/blob/master/refgenie/asset_build_packages.py) if you want to take a look

**Comment by nsheff** [comment]
So @k3yavi, I think @michal has added those 3 assets now for human, mouse and rat. So I think what we need from you is to 1) confirm that the recipes (and the results) are as you expect them to be; and 2) check the input files we're using (the txome and genome fasta files) to make sure this is the ones you're expecting to serve. If you want any more, either versions of the 3 species that we already have, or other species, you can add them directly.

**Comment by k3yavi** [comment]
Hi guys,
 
Thanks for the heads up. I was wondering is the memory issue related to SA is solved ? or we found a workaround ? As @nsheff mentioned, I checked the recipes [here](https://github.com/databio/refgenie/blob/master/refgenie/asset_build_packages.py#L220-L316) and it looks great to me. 
To verify the results I installed a fresh version of refgenie but somehow I am not able to pull either of  the SA versions of the salmon indices, are they online ? however the cdna only index looks great and produces the same results on our end as well.
Regarding the input files I am checking it [here](https://github.com/databio/databio_genomes/blob/master/asset_pep/recipe_inputs.csv#L4) and the annotation files (GTF) file source from ensembl looks right, I do see gencode version too and wondering do we already have assets build on top of gencode annotations ?
The source for cdna fasta from ensembl for hg19, mm10, rn6 looks great however I was wondering why are we using databio based version for `hg38_cdna-fasta` sample_name ?
Regarding the genome I think we might be using difference source, they might be the same files but I have to check; the regex for downloading from ensembl we generally use is the following:
```
genome: ftp://ftp.ensembl.org/pub/release-97//fasta/SPECIES/dna/\*.dna.toplevel.fa.gz

transcriptome: ftp://ftp.ensembl.org/pub/release-97//fasta/SPECIES/cdna/\*.cdna.all.fa.gz

gtf: ftp://ftp.ensembl.org/pub/release-97//gtf/SPECIES/\*97.gtf.gz
```

**Comment by stolarczyk** [comment]
>  I am not able to pull either of the SA versions of the salmon indices, are they online ? 

They are not up yet. We cannot build the `salmon_sa_index` ones due to the issue discussed above. It's currently in cluster support people hands and they are still tracking the issue..
The other one (`salmon_partial_sa_index`) is currently in progress: recipe is ready and the assets are being built. So these should be up by the end of this week, maybe earlier. 

>  I do see gencode version too and wondering do we already have assets build on top of gencode annotations

Yes, we serve both gencode and ensembl-sourced GTFs. And as you can see [here](https://github.com/databio/databio_genomes/blob/master/asset_pep/recipe_inputs.csv#L38) the `salmon_partial_sa_index` recipe takes the gencode one as an input. Should we change that? 

>  I was wondering why are we using databio based version for hg38_cdna-fasta sample_name 

I don't know why I put this one there. But the `hg38_cdna/fasta` asset that we serve has been build based on our [old PEP organization](https://github.com/databio/databio_genomes/blob/master/build_pep/genomes.csv#L15), which lists the correct genome source. I'll update the new one. Thanks!

**Comment by k3yavi** [comment]
Thanks @michalstolarczyk  for confirming. No rush at all, but let me know once SA based ones are ready, I can sanity check them on my end as well. Regarding the annotations for `salmon_partial_sa_index`, my feeling is for the sake of consistency we should take all genome/txome/annotation from the same source either ensembl or gencode. May be there is no major affect at all but I am not sure if mixing them can create any inconsistencies ?

**Comment by nsheff** [comment]
> They are not up yet. We cannot build the salmon_sa_index ones due to the issue discussed above. It's currently in cluster support people hands and they are still tracking the issue..

We were able to build it locally though, can we just put that up for now?

**Comment by k3yavi** [comment]
@nsheff  I thought the above question was not for me, was it ? If it was then I think developing index on any linux environment is fine, as long as it's not too much work for you guys.

**Comment by nsheff** [comment]
> I thought the above question was not for me

Right, it was for Michal, and he came by my office and we talked in person about it. The cluster problem still hasn't been solved but I'm now set up the build the assets locally which should be done tomorrow.

**Comment by k3yavi** [comment]
Thanks for confirming Nathan, I look forward to it .

**Comment by stolarczyk** [comment]
@k3yavi, we're now hosting `salmon_sa_index` and `salmon_partial_sa_index` assets at http://refgenomes.databio.org/ for the organisms mentioned above. Could you check whether they are ok from your perspective?

**Comment by k3yavi** [comment]
Hi @michalstolarczyk ,

Sorry for the late response. I checked all the indices, they look great. Thanks for making this all work so beautifully into the refgenie framework. I'll close this issue now !
